ex/loqui/lib/loqui/server.ex

```diff
@@ -0,0 +1,136 @@
+defmodule Loqui.Server do
+
+  defmodule Http do
+    defmodule State do
+      defstruct [:handler, :path, :transport, :socket, :buffer, :handler_opts]
+
+      def new(socket, transport, opts) do
+        %__MODULE__{
+          path: Keyword.fetch!(opts, :loqui_path),
+          handler: Keyword.fetch!(opts, :handler),
+          socket: socket,
+          transport: transport,
+          buffer: "",
+          handler_opts: Keyword.get(opts, :handler_opts, %{})
+        }
+      end
+    end
+
+    require Logger
+    alias Loqui.RanchProtocol
+
+    @behaviour :ranch_protocol
+    @max_payload 1024 * 100
+
+    def start_link(ref, socket, transport, opts) do
+      pid = spawn_link(__MODULE__, :init, [{ref, socket, transport, opts}])
+      {:ok, pid}
+    end
+
+    def init({ref, socket, transport, opts}) do
+      state = State.new(socket, transport, opts)
+      transport_opts = opts
+        |> Keyword.get(:transport_opts, [])
+        |> Keyword.put(:active, true)
+
+      # This will fail if the path hasn't been set.
+      _loqui_path = Keyword.fetch!(opts, :loqui_path)
+      :ok = :ranch.accept_ack(ref)
+      transport.setopts(socket, transport_opts)
+
+      loop(state)
+    end
+
+    def loop(%{socket: sock} = state) do
+      receive do
+        {:tcp, ^sock, data} ->
+          case handle_tcp_data(data, state) do
+            :ok ->
+              exit(:normal)
+
+            {:error, {:not_complete, request}} ->
+              loop(%State{state | buffer: request})
+
+            {:error, reason} ->
+              Logger.error("Client #{ip_address(sock)} caused #{inspect reason}. Exiting.")
+              exit(reason)
+          end
+
+        {:tcp_error, ^sock, reason} ->
+          Logger.warn("TCP error #{inspect reason} from client #{ip_address(sock)}. Closing")
+          exit(reason)
+
+        {:tcp_closed, ^sock} ->
+          Logger.info("Client #{ip_address(sock)} closed.")
+          exit(:normal)
+      end
+    end
+
+    defp handle_tcp_data(extra_data, %{socket: sock, path: loqui_path, transport: transport, buffer: buffer} = state) do
+      with {:ok, ^loqui_path, {headers, _}} <- try_parse_request(buffer <> extra_data),
+           "loqui" <- :proplists.get_value("upgrade", headers) do
+
+        response = :cow_http.response(101, :"HTTP/1.1", [{"upgrade", "loqui"}])
+        transport.send(sock, response)
+        RanchProtocol.upgrade(sock, transport, state.handler, state.handler_opts)
+        :ok
```

Comment:
Should this function return the result of `RanchProtocol.upgrade`? I think we're going to swallow some errors if we don't.

Reply:
ok